src/System.Device.Gpio/System/Device/I2c/I2cController.cs

```diff
@@ -0,0 +1,218 @@
+        /// <param name="shouldDispose">Determines if the device should be disposed by a controller or binding.</param>
+        public void OpenDevice(I2cConnectionSettings settings, bool shouldDispose = true)
+        {
+            int busId = settings.BusId;
+            int address = settings.DeviceAddress;
+
+            if (_openDevices.ContainsKey((busId, address)))
+            {
+                throw new InvalidOperationException($"The specified bus ID ({busId}) and address ({address}) is already used with another I2C device.");
+            }
+
+            I2cDevice device = I2cController.Create(new I2cConnectionSettings(busId, address));
+            device.ShouldDispose = shouldDispose;
+            _openDevices.Add((busId, address), device);
+        }
+
+        /// <summary>
+        /// Opens an I2C device in order for it to be ready to use with the controller.
+        /// </summary>
+        /// <param name="device">The I2C device to open.</param>
+        /// <param name="shouldDispose">Determines if the device should be disposed by a controller or binding.</param>
+        public void OpenDevice(I2cDevice device, bool shouldDispose = true)
+        {
+            int busId = device.ConnectionSettings.BusId;
+
+            foreach (I2cDevice openDevice in _openDevices.Values)
+            {
+                if (busId == openDevice.ConnectionSettings.BusId &&
+                    !device.GetType().Equals(openDevice.GetType()))
+                {
+                    throw new InvalidOperationException("All devices on an I2C bus must be of the same type.");
+                }
+            }
+
+            OpenDevice(device.ConnectionSettings, shouldDispose);
+        }
+
+        /// <summary>
+        /// Closes an open I2C device based on specified settings.
+        /// </summary>
+        /// <param name="settings">The connection settings of a device on an I2C bus.</param>
+        public void CloseDevice(I2cConnectionSettings settings)
+        {
+            int busId = settings.BusId;
+            int address = settings.DeviceAddress;
+
+            I2cDevice device = GetDevice(busId, address);
+            _openDevices.Remove((busId, address));
+
+            if (device.ShouldDispose)
+            {
+                device.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Closes an open I2C device.
+        /// </summary>
+        /// <param name="device">The I2C device to close.</param>
+        public void CloseDevice(I2cDevice device)
+        {
+            int busId = device.ConnectionSettings.BusId;
+            int address = device.ConnectionSettings.DeviceAddress;
+
+            _openDevices.Remove((busId, address));
+
+            if (device.ShouldDispose)
+            {
+                device.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Removes all I2C devices.
+        /// </summary>
+        public void ClearDevices() => Dispose(true);
```

Comment:
This seems redundant considering it does the same what Dispose

Reply:
Not included any longer.